code/modules/mining/mine_structures.dm

```diff
@@ -75,14 +97,49 @@
 		return
 	//Store our light's vars in here
 	flickering = 1
-	spawn(0)
+	spawn()
 		for(var/i = 0; i < amount; i++)
+			if(!lantern)
+				update_brightness()
+				break
 			set_light(0)
-			spawn(rand(5, 15))
-				set_light(6, 2, LIGHT_COLOR_TUNGSTEN)
-		set_light(6, 2, LIGHT_COLOR_TUNGSTEN)
+			sleep(rand(5, 15))
+			update_brightness()
+
 	flickering = 0
 
+/obj/structure/hanging_lantern/proc/update_brightness()
+
+	if(lantern)
+		light_range = lantern.light_range
+		light_power = lantern.light_power
+		light_color = lantern.light_color
+	else
+		light_range = 0
+		light_power = 0
+		light_color = 0
+
+	set_light(light_range, light_power, light_color)
+
+/obj/structure/hanging_lantern/verb/toggle_lantern()
```

Comment:
You seriously put ghost_read and ghost_write = 0 but you didn't put any protections in this verb?